# Aljabar Matriks

Matriks adalah suatu susunan bilangan real atau bilangan kompleks (atau elemen-elemen) yang disusun dalam baris dan kolom sehingga membentuk jajaran persegi panjang. Suatu matriks diberi nama dengan menggunakan huruf kapital seperti A, B, C, dan seterusnya, sedangkan anggotanya dinyatakan dengan huruf kecil.

Kita bisa membuat matriks di python dengan fungsi np.array yang nilainya bisa diisi sendiri

In [1]:
import numpy as np
A = np.array ([[1, 3, 4], [2, 1, 5]])
B = np.array ([[3, 0, 4], [-1, -2, -1]])
C = np.array ([[2, 8, 7], [6, 5, 2]])
D = np.array ([[12, 4]])
print(A)
print(B)
print(C)
print(D)

[[1 3 4]
 [2 1 5]]
[[ 3  0  4]
 [-1 -2 -1]]
[[2 8 7]
 [6 5 2]]
[[12  4]]


Atau kita bisa membuat matriks dengan fungsi random dengan ukuran yang bisa ditentukan sendiri ,contoh matriks ukuran 3x3

In [2]:
A = np.random.randint(10, size=(3,3))
print(A)

[[5 9 0]
 [1 4 7]
 [7 0 7]]


## Operasi Bilangan Matriks

### Penjumlahan Matriks

Penjumlahan Matriks pada dasarnya sama dengan penjumlahan bilangan, namun dengan salah satu syarat yaitu matriks yang dijumlahkan memiliki ordo yang sama. contohnya matriks $A$ dengan ordo 2x2 dijumlahkan dengan matriks $B$ dengan ordo 2x2 atau matriks $C$ dengan ordo 3x3 dijumlahkan dengan matriks $D$ dengan ordo 3x3    

Berikut Sifat sifat Penjumlahan Matriks


1. Sifat komutatif.
Pada sifat ini berlaku A + B = B + A.

2. Sifat asosiatif.
Pada sifat ini berlaku (A + B) + C = A +(B + C)

3. Matriks nol.
Pada sifat ini berlaku A + 0 = A.

4. Jika dijumlahkan dengan lawannya akan menghasilkan matriks nol, yaitu A + (-A) = 0. 


Contoh penjumlahan matriks

$\begin{bmatrix}
1 & 2\\
3 & 4
\end{bmatrix}$
+
$\begin{bmatrix}
2 & 4\\
6 & 8
\end{bmatrix}$
::
$\begin{bmatrix}
3 & 6\\
9 & 12
\end{bmatrix}$

### Perkalian Matriks

Perkalian pada Matriks terbagi menjadi 2, yaitu perkalian matriks dengan konstanta dan perkalian antarmatriks 

#### 1. Perkalian Matriks dengan Konstanta
Adalah suatu operasi yang mengalikan matriks dengan suatu nilai konstanta

Contoh perkalian Matriks dengan Konstanta:
Matriks M
$\begin{bmatrix}
1 & 2\\
3 & 4
\end{bmatrix}$
dikalikan dengan konstanta bernilai 5

In [3]:
M = np.array([[1,2], [3,4]])
print(M)
N = M * 5
print(N)

[[1 2]
 [3 4]]
[[ 5 10]
 [15 20]]


Hasilnya adalah Matriks N
$\begin{bmatrix}
5 & 10\\
15 & 20
\end{bmatrix}$

#### 2. Perkalian AntarMatriks

Adalah Operasi yang mengalikan matriks dengan matriks lainnya salah satu syarat perkalian antarmatriks adalah kolom matriks yang mengalikan sama dengan
baris matriks yang dikalikan

Contoh perkalian antarmatriks

In [4]:
#menentukan matriks yang akan dikalikan
A = np.random.randint(10, size=(2,3))
B = np.random.randint(10, size=(3,2))
print(A)
print(B)

[[5 4 6]
 [6 2 7]]
[[8 1]
 [4 2]
 [3 1]]


In [5]:
mA = [
    [4,2,1],
    [0,8,7],
]
mB = [
    [0,5],
    [2,2],
    [6,5],
]

print("Matriks A", mA, "\n")
print("Matriks B", mB, "\n")

print("Panjang baris dan kolom")

print("Matriks A")
print(len(mA),len(mA[0]),"\n")

print("Matriks B")
print(len(mB), len(mB[0]),"\n")

# x adalah panjang baris matriks A
x = len(mA)

# y adalah panjang kolom pertama matriks A
y = len(mA[0])

# z adalah panjang baris pertama matriks B
z = len(mB[0])

hasil = []
# perulangan yang terjadi sebanyak baris mA
for i in range(0, x):
    hasil1=[]
    #perulangan yang terjadi sebanyak kolom mB 
    for j in range(0, z):
        jumlah = 0
        for k in range(0, y):
            jumlah += mA[i][k] * mB[k][j]
        hasil1.append(jumlah)
    hasil.append(hasil1)
print("hasil dari perkalian matriks")
print(hasil,"\n")



Matriks A [[4, 2, 1], [0, 8, 7]] 

Matriks B [[0, 5], [2, 2], [6, 5]] 

Panjang baris dan kolom
Matriks A
2 3 

Matriks B
3 2 

hasil dari perkalian matriks
[[10, 29], [58, 51]] 



# Matriks Transpose
Matriks transpose matriks yang merupakan hasil dari membalikkan baris dan kolom matriks itu sendiri. Jadi, baris dari matriks transpose adalah kolom dari matriks itu sendiri dan kolom dari matriks transpose adalah baris dari matriks itu sendiri. Misalkan terdapat sebuah matriks A dengan ukuran $3 \times 2$, jadi transpose dari matriks A adalah matriks A transpose dengan ukuran $2 \times 3$   

Penggunaan matriks transpose pada python adalah sebagai berikut

In [6]:
A = np.array([[5, 4, 0],[1,8,3],[6,7,2]])
A_T = A.transpose()
print(A)
print('\n')
print(A_T)

[[5 4 0]
 [1 8 3]
 [6 7 2]]


[[5 1 6]
 [4 8 7]
 [0 3 2]]


Terdapat matriks A dengan ukuran $3 \times 3$ diatas. Cara menemukan transpose matriks A adalah dengan menggunakan fungsi A.transpose() pada python dan hasilnya seperti yang diatas

# Matriks Elementer
Matriks elementer adalah sebuah matriks yang berbentuk persegi yang dikenakan sebuah Operasi Baris Elementer (OBE). Dalam OBE, terdapat 3 operasi dasar. Maka dari itu, terdapat 3 cara untuk membuat matriks elementer, yaitu: 

1.Menukarkan dua baris pada matriks. \
 Dinotasikan dengan $R_i \leftrightarrow R_j$,

Contohnya adalah terdapat sebuah matriks A dengan ukuran $3 \times 3$ sebagai berikut \
$\begin {bmatrix}
2 & 0 & 1 \\
5 & 8 & 4 \\
1 & 0 & 2 
\end {bmatrix}$

Kita akan tukarkan posisi baris 1 dengan baris 3 dengan menggunakan operasi $A_1 \leftrightarrow A_3$ maka akan jadi seperti ini \
$\begin {bmatrix}
2 & 0 & 1 \\
5 & 8 & 4 \\
1 & 0 & 2 
\end {bmatrix}$
$\rightarrow$
$\begin {bmatrix}
1 & 0 & 2 \\
5 & 8 & 4 \\
2 & 0 & 1
\end {bmatrix}$

2.Mengalikan baris dengan konstanta, selama konstanta tidak sama dengan 0. \
 Dinotasikan dengan $kR_i \rightarrow R_i$


Contoh terdapat matriks $I_{2 \times 2}$ yang dimana akan dikenakan operasi $2R_2 \rightarrow R_2$ maka akan diperoleh hasil seperti berikut

$\begin {bmatrix}
1 & 0 \\
0 & 1
\end {bmatrix}$
$\rightarrow $
$\begin {bmatrix}
1 & 0 \\
0 & 2
\end {bmatrix}$

3. Menambahkan kelipatan suatu baris dengan baris yang lain. \
 Dinotasikan dengan $kR_i + R_j \rightarrow R_j$

contoh terdapat matriks $I_{3 \times 3}$ dengan dikenakan operasi $3R_2 + R_3 \rightarrow R_3$ maka diperoleh hasil seperti berikut

$\begin {bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1 
\end {bmatrix}$
$\rightarrow$
$\begin {bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 3 & 1 
\end {bmatrix}$

# Matriks Invers
Matriks invers adalah kebalikan dari suatu matriks. jadi jika dilambangkan suatu matriks adalah matriks $A$ maka kebalikan atau invers dari matriks $A$ adalah matriks $A^{-1}$. Matriks $A^{-1}$ dapat dikatakan invers dari matriks $A$ jika memenuhi syarat apabila matriks $A$ dikalikan dengan matriks $A^{-1}$ menghasilkan matriks $I$ yang mana matriks $I$ adalah matriks identitas

Untuk mencari matriks invers salah satu caranya adalah dengan menggunakan metode eliminasi Gauss-Jordan. Metode Gauss Jordan adalah adalah salah satu metode eliminasi yang digunakan untuk menyelesaikan sistem persamaan linear dan mendapatkan solusi yang diberikan dengan cara diubah menjadi augmented matriks atau matriks yang diperluas yang disederhanakan menjadi bentuk baris yang tereduksi.

Contoh cara mencari invers matriks menggunakan metode Gauss-Jordan adalah sebagai berikut

Diketahui matriks $A_{3 \times 3}$ sebagai berikut

$\begin {bmatrix}
2 & 2 & 1 \\
4 & 2 & 3 \\
1 & 3 & -1 
\end {bmatrix}$

In [7]:
A = np.array([[2,2,1],[4,2,3],[1,3,-1]])
print(A)

[[ 2  2  1]
 [ 4  2  3]
 [ 1  3 -1]]


In [8]:
def RowSwap(A,k,l):
# =============================================================================
#     A is a NumPy array.  RowSwap will return duplicate array with rows
#     k and l swapped.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp

    return B

def RowScale(A,k,scale):
# =============================================================================
#     A is a NumPy array.  RowScale will return duplicate array with the
#     entries of row k multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale

    return B

def RowAdd(A,k,l,scale):
# =============================================================================
#     A is a numpy array.  RowAdd will return duplicate array with row
#     l modifed.  The new values will be the old values of row l added to
#     the values of row k, multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[l][j] += B[k][j]*scale

    return B

Kita akan menggunakan fungsi **RowAdd**, **RowScale**, dan **RowSwap**

In [9]:
A1 = RowSwap(A,0,2)
print(A1)

[[ 1.  3. -1.]
 [ 4.  2.  3.]
 [ 2.  2.  1.]]


In [10]:
A2 = RowAdd(A1,0,1,-4)
print(A2)

[[  1.   3.  -1.]
 [  0. -10.   7.]
 [  2.   2.   1.]]


In [11]:
A3 = RowAdd(A2,0,2,-2)
print(A3)

[[  1.   3.  -1.]
 [  0. -10.   7.]
 [  0.  -4.   3.]]


In [12]:
A4 = RowSwap(A3,1,2)
print(A4)

[[  1.   3.  -1.]
 [  0.  -4.   3.]
 [  0. -10.   7.]]


In [13]:
A5 = RowAdd(A4,1,2,-2)
print(A5)

[[ 1.  3. -1.]
 [ 0. -4.  3.]
 [ 0. -2.  1.]]


In [14]:
A6 = RowAdd(A5,2,0,1)
print(A6)

[[ 1.  1.  0.]
 [ 0. -4.  3.]
 [ 0. -2.  1.]]


In [15]:
A7 = RowAdd(A6,2,1,-3)
print(A7)

[[ 1.  1.  0.]
 [ 0.  2.  0.]
 [ 0. -2.  1.]]


In [16]:
A8= RowAdd(A7,1,0,-0.5)
print(A8)

[[ 1.  0.  0.]
 [ 0.  2.  0.]
 [ 0. -2.  1.]]


In [17]:
A9 = RowScale(A8,1,1/2)
print(A9)

[[ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0. -2.  1.]]


In [18]:
A10 = RowAdd(A9,1,2,2)
print(A10)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Setelah menemukan matriks identitasnya, kita akan lakukan operasi yang sama pada matriks identitas

In [19]:
B = np.array([[1,0,0],[0,1,0],[0,0,1]])
print(B)

[[1 0 0]
 [0 1 0]
 [0 0 1]]


In [20]:
B1 = RowSwap(B,0,2)
print(B1)

[[0. 0. 1.]
 [0. 1. 0.]
 [1. 0. 0.]]


In [21]:
B2 = RowAdd(B1,0,1,-4)
print(B2)

[[ 0.  0.  1.]
 [ 0.  1. -4.]
 [ 1.  0.  0.]]


In [22]:
B3 = RowAdd(B2,0,2,-2)
print(B3)

[[ 0.  0.  1.]
 [ 0.  1. -4.]
 [ 1.  0. -2.]]


In [23]:
B4 = RowSwap(B3,1,2)
print(B4)

[[ 0.  0.  1.]
 [ 1.  0. -2.]
 [ 0.  1. -4.]]


In [24]:
B5 = RowAdd(B4,1,2,-2)
print(B5)

[[ 0.  0.  1.]
 [ 1.  0. -2.]
 [-2.  1.  0.]]


In [25]:
B6 = RowAdd(B5,2,0,1)
print(B6)

[[-2.  1.  1.]
 [ 1.  0. -2.]
 [-2.  1.  0.]]


In [26]:
B7 = RowAdd(B6,2,1,-3)
print(B7)

[[-2.  1.  1.]
 [ 7. -3. -2.]
 [-2.  1.  0.]]


In [27]:
B8= RowAdd(B7,1,0,-0.5)
print(B8)

[[-5.5  2.5  2. ]
 [ 7.  -3.  -2. ]
 [-2.   1.   0. ]]


In [28]:
B9 = RowScale(B8,1,1/2)
print(B9)


[[-5.5  2.5  2. ]
 [ 3.5 -1.5 -1. ]
 [-2.   1.   0. ]]


In [29]:
B10 = RowAdd(B9,1,2,2)
print(B10)

[[-5.5  2.5  2. ]
 [ 3.5 -1.5 -1. ]
 [ 5.  -2.  -2. ]]


Dengan ini, telah kita peroleh matriks invers dari matriks A adalah 

$\begin{bmatrix}
-5.5 & 2.5 & 2 \\
3.5 & -1.5 & -1 \\
5 & -2 & -2
\end {bmatrix}$

Referensi 
1. https://www.ruangguru.com/blog/mengenal-matriks-dalam-matematika-pengertian-jenis-dan-transpose
2. https://gdscuniversitasmulia.medium.com/eliminasi-gauss-jordan-79b87e67befd
3. https://www.profematika.com/definisi-matriks-elementer-dan-sifatnya/#google_vignette
4. https://aljabarlinear.mipa.ugm.ac.id/soal/uts-uas/eleminasi-gauss-jordan/
5. https://akupintar.id/info-pintar/-/blogs/matriks-pengertian-operasi-determinan-invers-dan-contoh-soal